### PR TITLE
Introduce request pools

### PR DESCRIPTION
**Pull request by abrooks98**
## Author Checklist
* [ ] Reference appropriate issues (with "Fixes" or "See" as appropriate)
* [ ] Remove xfail from the test suite when fixing a test
* [ ] Commits are self-contained and do not do two things at once
* [ ] Commit message is of the form: `module: short description` and follows [good practice](https://chris.beams.io/posts/git-commit/)
* [x] Passes whitespace checkers
* [x] Passes warning tests
* [x] Passes all tests
* [ ] Add comments such that someone without knowledge of the code could understand
* [ ] Add Devel Docs in the `doc/` directory for any new code design

### CODE REVIEW

**Comment by abrooks98** [comment]
test:jenkins/ch4/most

File: src/include/mpir_objects.h [hunk 1]
```diff
@@ -174,10 +174,16 @@ const char *MPIR_Handle_get_kind_str(int kind);
...
+#define HANDLE_POOL_MASK 0x03f00000
+#define HANDLE_POOL_SHIFT 20
+#define HANDLE_POOL_INDEX(a) (((a) & HANDLE_POOL_MASK) >> HANDLE_POOL_SHIFT)
+#define HANDLE_NUM_POOLS ((HANDLE_POOL_MASK >> HANDLE_POOL_SHIFT) + 1)
+
 /* For indirect, the remainder of the handle has a block and index within that
  * block */
 #define HANDLE_INDIRECT_SHIFT 12
 #define HANDLE_BLOCK(a) (((a)& 0x03FFF000) >> HANDLE_INDIRECT_SHIFT)
+#define HANDLE_PBLOCK(a) (((a)& 0x0003F000) >> HANDLE_INDIRECT_SHIFT)
```
Comment (hzhou) [inline src/include/mpir_objects.h]: Shouldn't this be `0x000FF000`?

File: src/include/mpir_objects.h [hunk 2]
```diff
@@ -423,12 +438,21 @@ typedef struct MPIR_Object_alloc_t {
...
     int direct_size;            /* Size of direct block */
+    int indirect_num_blocks;    /* Max. number of indirect blocks */
+    int indirect_num_indices;   /* Number of objects in one indirect block */
+    int pool_idx;               /* Pool index within kind */
 } MPIR_Object_alloc_t;
 static inline void *MPIR_Handle_obj_alloc(MPIR_Object_alloc_t *);
 static inline void *MPIR_Handle_obj_alloc_unsafe(MPIR_Object_alloc_t *);
 static inline void MPIR_Handle_obj_free(MPIR_Object_alloc_t *, void *);
-static inline void *MPIR_Handle_get_ptr_indirect(int, MPIR_Object_alloc_t *);
+MPL_STATIC_INLINE_PREFIX void *MPIR_Handle_get_ptr_indirect(int, MPIR_Object_alloc_t *);
```
Comment (hzhou) [inline src/include/mpir_objects.h]: Rather than doing it partially here, I would do a global refactor for it.

File: src/include/mpir_objects.h [hunk 3]
```diff
@@ -423,12 +438,21 @@ typedef struct MPIR_Object_alloc_t {
...
+MPL_STATIC_INLINE_PREFIX void *MPIR_Handle_get_ptr_indirect_generic(int, MPIR_Object_alloc_t *,
```
Comment (hzhou) [inline src/include/mpir_objects.h]: Long but uninformative names :)

File: src/include/mpir_objects.h [hunk 4]
```diff
@@ -423,12 +438,21 @@ typedef struct MPIR_Object_alloc_t {
...
+                                                                    int);
 
+void MPIR_Object_alloc_populate(MPIR_Object_alloc_t * objmem,
+                                MPII_Object_kind kind,
+                                int size, void *direct, int direct_size,
+                                int indirect_num_blocks, int indirect_num_indices, int num_pools);
```
Comment (hzhou) [inline src/include/mpir_objects.h]: Adding a `num_pools` parameters is called `_populate`?

**Comment by hzhou** [comment]
Superseded by #4285

**Pull request closed without merging**